xcodeproj/internal/xcode_targets.bzl

```
@@ -267,74 +282,120 @@ def _to_xcode_target_product(product):
         _additional_files = product.framework_files,
     )
 
-def _merge_xcode_target(*, src, dest):
+def _merge_xcode_target(*, src_swift, src_non_swift, dest):
     """Creates a new `xcode_target` by merging the values of `src` into `dest`.
 
     Args:
-        src: The `xcode_target` to merge into `dest`.
+        src_swift: The `xcode_target` to merge into `dest` that `is_swift`.
+        src_non_swift: The `xcode_target` to merge into `dest` that is not
+            `is_swift`.
         dest: The `xcode_target` being merged into.
 
     Returns:
         A value as returned by `xcode_targets.make`.
     """
 
-    build_settings = dict(structs.to_dict(src._build_settings))
+    build_settings = {}
+
+    if src_swift:
+        build_settings = dicts.add(
+            structs.to_dict(src_swift._build_settings),
+            build_settings,
+        )
+    if src_non_swift:
+        build_settings = dicts.add(
+            structs.to_dict(src_non_swift._build_settings),
+            build_settings,
+        )
+
+    # This is intentionally last since it's the highest-level target that should
+    # override some build settings from the other target(s).
     build_settings = dicts.add(
         structs.to_dict(dest._build_settings),
         build_settings,
     )
 
+    srcs = []
+    transitive_dependencies = [dest._dependencies]
+    platform = None
+    c_params = dest._c_params
+    cxx_params = dest._cxx_params
+    swift_params = dest._swift_params
+    c_has_fortify_source = dest._c_has_fortify_source
+    cxx_has_fortify_source = dest._cxx_has_fortify_source
+    swiftmodules = []
+    modulemaps = ()
+    if src_swift:
+        srcs.append(src_swift)
+        transitive_dependencies.append(src_swift._dependencies)
+        swift_params = src_swift._swift_params or dest._swift_params
+        platform = src_swift.platform
+        swiftmodules = src_swift._swiftmodules
+        modulemaps = src_swift._modulemaps
+    if src_non_swift:
+        srcs.append(src_non_swift)
+        transitive_dependencies.append(src_non_swift._dependencies)
+        c_params = src_non_swift._c_params or dest._c_params
+        cxx_params = src_non_swift._cxx_params or dest._cxx_params
+        c_has_fortify_source = src_non_swift._c_has_fortify_source or dest._c_has_fortify_source
+        cxx_has_fortify_source = src_non_swift._cxx_has_fortify_source or dest._cxx_has_fortify_source
```

Since you set these to `dest` above, we can use that value in the rhs.

```suggestion
        c_params = src_non_swift._c_params or c_params
        cxx_params = src_non_swift._cxx_params or cxx_params
        c_has_fortify_source = src_non_swift._c_has_fortify_source or c_has_fortify_source
        cxx_has_fortify_source = src_non_swift._cxx_has_fortify_source or cxx_has_fortify_source
```